

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Cooler module &mdash; FAN-C 0.9.10-beta documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="FAN-C Python architecture modules" href="../architecture.html" />
    <link rel="prev" title="Juicer module" href="juicer.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> FAN-C
          

          
          </a>

          
            
            
              <div class="version">
                0.9.10
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started with FAN-C</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../fanc.html">fanc: command line tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../fancplot.html">fancplot: plotting from the command line</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../compatibility.html">Working with Juicer and Cooler files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../interfaces.html">Interfaces</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../generate.html">Generate Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analyse.html">Analyse Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../plot.html">Plotting (API)</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../modules.html">Reference</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../init.html">init</a></li>
<li class="toctree-l3"><a class="reference internal" href="../map.html">map</a></li>
<li class="toctree-l3"><a class="reference internal" href="../regions.html">regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pairs.html">pairs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../matrix.html">matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hic.html">hic</a></li>
<li class="toctree-l3"><a class="reference internal" href="../peaks.html">peaks</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../compatibility.html">compatibility</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="juicer.html">Juicer</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">Cooler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../architecture.html">architecture</a></li>
<li class="toctree-l3"><a class="reference internal" href="../plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools.html">tools</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">FAN-C</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../api.html">fanc Python module</a> &raquo;</li>
        
          <li><a href="../../modules.html">FAN-C Python modules</a> &raquo;</li>
        
          <li><a href="../compatibility.html">FAN-C compatibility modules</a> &raquo;</li>
        
      <li>Cooler module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../_sources/api/modules/compatibility/cooler.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-fanc.compatibility.cooler">
<span id="cooler-module"></span><h1>Cooler module<a class="headerlink" href="#module-fanc.compatibility.cooler" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="fanc.compatibility.cooler.CoolerHic">
<em class="property">class </em><code class="sig-prename descclassname">fanc.compatibility.cooler.</code><code class="sig-name descname">CoolerHic</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="../matrix.html#fanc.matrix.RegionMatrixContainer" title="fanc.matrix.RegionMatrixContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.RegionMatrixContainer</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">cooler.api.Cooler</span></code></p>
<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_contact">
<code class="sig-name descname">add_contact</code><span class="sig-paren">(</span><em class="sig-param">contact</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_contact" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contact</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_contacts">
<code class="sig-name descname">add_contacts</code><span class="sig-paren">(</span><em class="sig-param">contacts</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <code class="xref py py-func docutils literal notranslate"><span class="pre">add_edges()</span></code></p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_edge">
<code class="sig-name descname">add_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">check_nodes_exist=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an edge / contact between two regions to this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code>, dict with at least the
attributes source and sink, optionally weight,
or a list of length 2 (source, sink) or 3
(source, sink, weight).</p></li>
<li><p><strong>check_nodes_exist</strong> – Make sure that there are nodes
that match source and sink indexes</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_edge_from_dict">
<code class="sig-name descname">add_edge_from_dict</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_edge_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from dict input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – dict with at least the keys “source”
and “sink”. Additional keys will be loaded
as edge attributes</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_edge_from_edge">
<code class="sig-name descname">add_edge_from_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_edge_from_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_edge_from_list">
<code class="sig-name descname">add_edge_from_list</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_edge_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from list or tuple input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – List or tuple. Should be of length 2
(source, sink) or 3 (source, sink, weight)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_edge_simple">
<code class="sig-name descname">add_edge_simple</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">weight=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_edge_simple" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> – Source region index</p></li>
<li><p><strong>sink</strong> – Sink region index</p></li>
<li><p><strong>weight</strong> – Weight of the edge</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_edges">
<code class="sig-name descname">add_edges</code><span class="sig-paren">(</span><em class="sig-param">edges</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk-add edges from a list.</p>
<p>List items can be any of the supported edge types,
list, tuple, dict, or <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code>. Repeatedly
calls <code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code>, so
may be inefficient for large amounts of data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edges</strong> – List (or iterator) of edges. See
<code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code>
for details</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.bin_size">
<em class="property">property </em><code class="sig-name descname">bin_size</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.bins">
<code class="sig-name descname">bins</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin table selector</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Table selector</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.bins_to_distance">
<code class="sig-name descname">bins_to_distance</code><span class="sig-paren">(</span><em class="sig-param">bins</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.bins_to_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.binsize">
<em class="property">property </em><code class="sig-name descname">binsize</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.binsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Resolution in base pairs if uniform else None</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.chromnames">
<em class="property">property </em><code class="sig-name descname">chromnames</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.chromnames" title="Permalink to this definition">¶</a></dt>
<dd><p>List of reference sequence names</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.chromosome_bins">
<em class="property">property </em><code class="sig-name descname">chromosome_bins</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.chromosome_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.chroms">
<code class="sig-name descname">chroms</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Chromosome table selector</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Table selector</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.chromsizes">
<em class="property">property </em><code class="sig-name descname">chromsizes</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.chromsizes" title="Permalink to this definition">¶</a></dt>
<dd><p>Ordered mapping of reference sequences to their lengths in bp</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><em class="sig-param">remove_tmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close this Juicer file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.distance_to_bins">
<code class="sig-name descname">distance_to_bins</code><span class="sig-paren">(</span><em class="sig-param">distance</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.distance_to_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.edge_data">
<code class="sig-name descname">edge_data</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.edge_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over specific edge attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute, e.g. “weight”</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator over edge attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.edge_subset">
<code class="sig-name descname">edge_subset</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.edge_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subset of edges.</p>
<p>This is an alias for <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generator (<code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.edges">
<em class="property">property </em><code class="sig-name descname">edges</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over contacts / edges.</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code> is the central function of
<code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code>. Here, we will use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> implementation for demonstration purposes,
but the usage is exactly the same for all compatible
objects implementing <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code>, including
<a class="reference internal" href="juicer.html#fanc.compatibility.juicer.JuicerHic" title="fanc.compatibility.juicer.JuicerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">JuicerHic</span></code></a> and
<a class="reference internal" href="#fanc.compatibility.cooler.CoolerHic" title="fanc.compatibility.cooler.CoolerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">CoolerHic</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># file from FAN-C examples</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can easily find the number of edges in the sample
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span>  <span class="c1"># 8695</span>
</pre></div>
</div>
<p>When used in an iterator context, <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>
iterates over all edges in the <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Calling <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code> as a method has the
same effect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># note the &#39;()&#39;</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Rather than iterate over all edges in the object, we can select only a subset.
If the key is a string or a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, all non-zero edges connecting
the region described by the key to any other region are returned. If the key is a
tuple of strings or <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, only edges between the two regions
are returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all edges between chromosome 19</span>
<span class="c1"># and any other region:</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>

<span class="c1"># select all edges that are only on</span>
<span class="c1"># chromosome 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 90--116; bias: 0.00021173151730025176; sink_node: chr19:37000001-38000000; source_node: chr19:11000001-12000000; weight: 0.009104455243910825</span>
    <span class="c1"># 135--135; bias: 0.00018003890596887822; sink_node: chr19:56000001-57000000; source_node: chr19:56000001-57000000; weight: 0.10028167062466517</span>
    <span class="c1"># 123--123; bias: 0.00011063368998965993; sink_node: chr19:44000001-45000000; source_node: chr19:44000001-45000000; weight: 0.1386240135570439</span>
    <span class="c1"># 92--93; bias: 0.00040851066434864896; sink_node: chr19:14000001-15000000; source_node: chr19:13000001-14000000; weight: 0.10090213409411629</span>
    <span class="c1"># ...</span>

<span class="c1"># select inter-chromosomal edges</span>
<span class="c1"># between chromosomes 18 and 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>By default, <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code> will retrieve all edge attributes,
which can be slow when iterating over a lot of edges. This is why all file-based FAN-C
<code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code> objects support lazy loading, where attributes
are only read on demand.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span> <span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42 42 0.12291311562018173 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #0&gt;</span>
    <span class="c1"># 24 28 0.025205961072838057 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #1&gt;</span>
    <span class="c1"># 5 76 0.00961709840049876 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #2&gt;</span>
    <span class="c1"># 66 68 0.03876763316345468 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #3&gt;</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The lazy iterator reuses the <code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code> object in every iteration,
and overwrites the <code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code> attributes. Therefore <strong>do not</strong> use
lazy iterators if you need to store edge objects for later access.
For example, the following code works as expected
<code class="code docutils literal notranslate"><span class="pre">list(hic.edges())</span></code>, with all <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> objects stored in the
list, while this code <code class="code docutils literal notranslate"><span class="pre">list(hic.edges(lazy=True))</span></code>
will result in a list of identical <code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code> objects. Always ensure
you do all edge processing in the loop when working with lazy iterators!</p>
</div>
<p>When working with normalised contact frequencies, such as obtained through
matrix balancing in the example above, <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>
automatically returns normalised edge weights. In addition, the <code class="code docutils literal notranslate"><span class="pre">bias</span></code>
attribute will (typically) have a value different from 1.</p>
<p>When you are interested in the raw contact frequency, use the <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>
parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">)</span>
    <span class="c1"># 42 42 2120.0</span>
    <span class="c1"># 24 28 388.0</span>
    <span class="c1"># 5 76 94.0</span>
    <span class="c1"># 66 68 470.0</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>You can also choose to omit all intra- or inter-chromosomal edges using
<code class="code docutils literal notranslate"><span class="pre">intra_chromosomal=False</span></code> or <code class="code docutils literal notranslate"><span class="pre">inter_chromosomal=False</span></code>, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Iterator over <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> or equivalent.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.edges_dict">
<code class="sig-name descname">edges_dict</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.edges_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Edges iterator with access by bracket notation.</p>
<p>This iterator <strong>always</strong> returns unnormalised edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict or dict-like iterator</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.expected_values">
<code class="sig-name descname">expected_values</code><span class="sig-paren">(</span><em class="sig-param">selected_chromosome=None</em>, <em class="sig-param">norm=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.expected_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the expected values for genomic contacts at all distances.</p>
<p>This calculates the expected values between genomic regions
separated by a specific distance. Expected values are calculated
as the average weight of edges between region pairs with the same
genomic separation, taking into account unmappable regions.</p>
<p>It will return a tuple with three values: a list of genome-wide
intra-chromosomal expected values (list index corresponds to number
of separating bins), a dict with chromosome names as keys and
intra-chromosomal expected values specific to each chromosome, and
a float for inter-chromosomal expected value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_chromosome</strong> – (optional) Chromosome name. If provided,
will only return expected values for this
chromosome.</p></li>
<li><p><strong>norm</strong> – If False, will calculate the expected values on the
unnormalised matrix.</p></li>
<li><p><strong>args</strong> – Not used in this context</p></li>
<li><p><strong>kwargs</strong> – Not used in this context</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of intra-chromosomal expected values,
dict of intra-chromosomal expected values by chromosome,
inter-chromosomal expected value</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.expected_values_and_marginals">
<code class="sig-name descname">expected_values_and_marginals</code><span class="sig-paren">(</span><em class="sig-param">selected_chromosome=None</em>, <em class="sig-param">norm=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.expected_values_and_marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the expected values for genomic contacts at all distances
and the whole matrix marginals.</p>
<p>This calculates the expected values between genomic regions
separated by a specific distance. Expected values are calculated
as the average weight of edges between region pairs with the same
genomic separation, taking into account unmappable regions.</p>
<p>It will return a tuple with three values: a list of genome-wide
intra-chromosomal expected values (list index corresponds to number
of separating bins), a dict with chromosome names as keys and
intra-chromosomal expected values specific to each chromosome, and
a float for inter-chromosomal expected value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_chromosome</strong> – (optional) Chromosome name. If provided,
will only return expected values for this
chromosome.</p></li>
<li><p><strong>norm</strong> – If False, will calculate the expected values on the
unnormalised matrix.</p></li>
<li><p><strong>args</strong> – Not used in this context</p></li>
<li><p><strong>kwargs</strong> – Not used in this context</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of intra-chromosomal expected values,
dict of intra-chromosomal expected values by chromosome,
inter-chromosomal expected value</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.extent">
<code class="sig-name descname">extent</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.extent" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin IDs containing the left and right ends of a genomic region</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> (<em>str</em><em> or </em><em>tuple</em>) – Genomic range</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>2-tuple of ints</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">extent</span><span class="p">(</span><span class="s1">&#39;chr3&#39;</span><span class="p">)</span>  
<span class="go">(1311, 2131)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.info">
<em class="property">property </em><code class="sig-name descname">info</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.info" title="Permalink to this definition">¶</a></dt>
<dd><p>File information and metadata</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.mappable">
<code class="sig-name descname">mappable</code><span class="sig-paren">(</span><em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.mappable" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mappability vector of this matrix.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.marginals">
<code class="sig-name descname">marginals</code><span class="sig-paren">(</span><em class="sig-param">masked=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the marginals vector of this Hic matrix.</p>
<p>Sums up all contacts for each bin of the Hi-C matrix.
Unmappable regoins will be masked in the returned vector unless
the <code class="code docutils literal notranslate"><span class="pre">masked</span></code> parameter is set to <code class="code docutils literal notranslate"><span class="pre">False</span></code>.</p>
<p>By default, corrected matrix entries are summed up.
To get uncorrected matrix marginals use <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>.
Generally, all parameters accepted by <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>
are supported.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>masked</strong> – Use a numpy masked array to mask entries
corresponding to unmappable regions</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.matrix">
<code class="sig-name descname">matrix</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">log=False</em>, <em class="sig-param">default_value=None</em>, <em class="sig-param">mask=True</em>, <em class="sig-param">log_base=2</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Assemble a <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code> from region pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Matrix selector. See <a class="reference internal" href="../matrix.html#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
for all supported key types</p></li>
<li><p><strong>log</strong> – If True, log-transform the matrix entries. Also see log_base</p></li>
<li><p><strong>log_base</strong> – Base of the log transformation. Default: 2; only used when
log=True</p></li>
<li><p><strong>default_value</strong> – (optional) set the default value of matrix entries
that have no associated edge/contact</p></li>
<li><p><strong>mask</strong> – If False, do not mask unmappable regions</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<a class="reference internal" href="../matrix.html#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries"><code class="xref py py-func docutils literal notranslate"><span class="pre">regions_and_matrix_entries()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<a class="reference internal" href="../matrix.html#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries"><code class="xref py py-func docutils literal notranslate"><span class="pre">regions_and_matrix_entries()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="../matrix.html#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">pairs</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge two or more <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></p></li>
<li><p><strong>args</strong> – Positional arguments passed to constructor of this
class</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to constructor of this
class</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.offset">
<code class="sig-name descname">offset</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.offset" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin ID containing the left end of a genomic region</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> (<em>str</em><em> or </em><em>tuple</em>) – Genomic range</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">offset</span><span class="p">(</span><span class="s1">&#39;chr3&#39;</span><span class="p">)</span>  
<span class="go">1311</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.open">
<code class="sig-name descname">open</code><span class="sig-paren">(</span><em class="sig-param">mode='r'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.open" title="Permalink to this definition">¶</a></dt>
<dd><p>Open the HDF5 group containing the Cooler with <code class="xref py py-mod docutils literal notranslate"><span class="pre">h5py</span></code></p>
<p>Functions as a context manager. Any <code class="docutils literal notranslate"><span class="pre">open_kws</span></code> passed during
construction are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em><em> [</em><em>default: 'r'</em><em>]</em>) – <ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">'r'</span></code> (readonly)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'r+'</span></code> or <code class="docutils literal notranslate"><span class="pre">'a'</span></code> (read/write)</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>For other parameters, see <code class="xref py py-class docutils literal notranslate"><span class="pre">h5py.File</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.pixels">
<code class="sig-name descname">pixels</code><span class="sig-paren">(</span><em class="sig-param">join=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.pixels" title="Permalink to this definition">¶</a></dt>
<dd><p>Pixel table selector</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>join</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to expand bin ID columns into chrom, start, and end columns.
Default is <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Table selector</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.possible_contacts">
<code class="sig-name descname">possible_contacts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.possible_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the possible number of contacts in the genome.</p>
<p>This calculates the number of potential region pairs in
a genome for any possible separation distance, taking into
account the existence of unmappable regions.</p>
<p>It will calculate one number for inter-chromosomal pairs,
return a list with the number of possible pairs where the
list index corresponds to the number of bins separating two regions,
and a dictionary of lists for each chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>possible intra-chromosomal pairs,
possible intra-chromosomal pairs by chromosome,
possible inter-chromosomal pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.regions_and_edges">
<code class="sig-name descname">regions_and_edges</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.regions_and_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to regions and edges selected by key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge selector, see <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>args</strong> – Positional arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over edges</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.regions_and_matrix_entries">
<code class="sig-name descname">regions_and_matrix_entries</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">score_field=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.regions_and_matrix_entries" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to non-zero matrix entries and associated regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge key, see <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>oe</strong> – If True, will divide observed values by their expected value
at the given distance. False by default</p></li>
<li><p><strong>oe_per_chromosome</strong> – If True (default), will do a per-chromosome O/E
calculation rather than using the whole matrix
to obtain expected values</p></li>
<li><p><strong>score_field</strong> – (optional) any edge attribute that returns a number
can be specified here for filling the matrix. Usually
this is defined by the <code class="code docutils literal notranslate"><span class="pre">_default_score_field</span></code>
attribute of the matrix class.</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over (i, j, weight) tuples</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.regions_identical">
<em class="property">static </em><code class="sig-name descname">regions_identical</code><span class="sig-paren">(</span><em class="sig-param">pairs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.regions_identical" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the regions in all objects in the list are identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
objects</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if chromosome, start, and end are identical between
all regions in the same list positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.scaling_factor">
<code class="sig-name descname">scaling_factor</code><span class="sig-paren">(</span><em class="sig-param">matrix</em>, <em class="sig-param">weight_column=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.scaling_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the scaling factor to another matrix.</p>
<p>Calculates the ratio between the number of contacts in
this Hic object to the number of contacts in another
Hic object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object</p></li>
<li><p><strong>weight_column</strong> – Name of the column to calculate the scaling factor on</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.storage_mode">
<em class="property">property </em><code class="sig-name descname">storage_mode</code><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.storage_mode" title="Permalink to this definition">¶</a></dt>
<dd><p>Indicates whether ordinary sparse matrix encoding is used
(<code class="docutils literal notranslate"><span class="pre">&quot;square&quot;</span></code>) or whether a symmetric matrix is encoded by storing only
the upper triangular elements (<code class="docutils literal notranslate"><span class="pre">&quot;symmetric-upper&quot;</span></code>).</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.CoolerHic.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.CoolerHic.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion">
<em class="property">class </em><code class="sig-prename descclassname">fanc.compatibility.cooler.</code><code class="sig-name descname">LazyCoolerRegion</code><span class="sig-paren">(</span><em class="sig-param">series</em>, <em class="sig-param">ix=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">genomic_regions.regions.GenomicRegion</span></code></p>
<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.as_bed_line">
<code class="sig-name descname">as_bed_line</code><span class="sig-paren">(</span><em class="sig-param">score_field='score'</em>, <em class="sig-param">name_field='name'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.as_bed_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of this object as line in a BED file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>score_field</strong> – name of the attribute to be used in the
‘score’ field of the BED line</p></li>
<li><p><strong>name_field</strong> – name of the attribute to be used in the
‘name’ field of the BED line</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.as_gff_line">
<code class="sig-name descname">as_gff_line</code><span class="sig-paren">(</span><em class="sig-param">source_field='source'</em>, <em class="sig-param">feature_field='feature'</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">frame_field='frame'</em>, <em class="sig-param">float_format='.2e'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.as_gff_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of this object as line in a GFF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source_field</strong> – name of the attribute to be used in the
‘source’ field of the GFF line</p></li>
<li><p><strong>feature_field</strong> – name of the attribute to be used in the
‘feature’ field of the GFF line</p></li>
<li><p><strong>score_field</strong> – name of the attribute to be used in the
‘score’ field of the GFF line</p></li>
<li><p><strong>frame_field</strong> – name of the attribute to be used in the
‘frame’ field of the GFF line</p></li>
<li><p><strong>float_format</strong> – Formatting string for the float fields</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.attributes">
<em class="property">property </em><code class="sig-name descname">attributes</code><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.attributes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all visible attributes of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<p>Returns all attribute names that do not start with an underscore.
:return: list of attribute names</p>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.center">
<em class="property">property </em><code class="sig-name descname">center</code><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.center" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the center coordinate of the <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the specified region is completely contained in this region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object or string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a (shallow) copy of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.expand">
<code class="sig-name descname">expand</code><span class="sig-paren">(</span><em class="sig-param">absolute=None</em>, <em class="sig-param">relative=None</em>, <em class="sig-param">absolute_left=0</em>, <em class="sig-param">absolute_right=0</em>, <em class="sig-param">relative_left=0.0</em>, <em class="sig-param">relative_right=0.0</em>, <em class="sig-param">copy=True</em>, <em class="sig-param">from_center=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.expand" title="Permalink to this definition">¶</a></dt>
<dd><p>Expand this region by a relative or an absolute amount.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>absolute</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on both
sides. New region start will be
&lt;old start - absolute&gt;, new region end
will be &lt;old end + absolute&gt;</p></li>
<li><p><strong>relative</strong> – Relative amount as fraction of region by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on both
sides. New region start will be
&lt;old start - relative*len(self)&gt;,
new region end will be
&lt;old end + relative*(len(self)&gt;</p></li>
<li><p><strong>absolute_left</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
left side</p></li>
<li><p><strong>absolute_right</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
right side</p></li>
<li><p><strong>relative_left</strong> – Relative amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
left side</p></li>
<li><p><strong>relative_right</strong> – Relative amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
right side</p></li>
<li><p><strong>copy</strong> – If True, return a copy of the original region,
if False will modify the existing region in place</p></li>
<li><p><strong>from_center</strong> – If True measures distance from center rather
than start and end of the old region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.five_prime">
<em class="property">property </em><code class="sig-name descname">five_prime</code><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.five_prime" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the position of the 5’ end of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
on the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.fix_chromosome">
<code class="sig-name descname">fix_chromosome</code><span class="sig-paren">(</span><em class="sig-param">copy=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.fix_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Change chromosome representation from chr&lt;NN&gt; to &lt;NN&gt; or vice versa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>copy</strong> – If True, make copy of region, otherwise
will modify existing region in place.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">region_string</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string into a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<p>This is a very useful convenience function to quickly
define a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object from a descriptor
string. Numbers can be abbreviated as ‘12k’, ‘1.5M’, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region_string</strong> – A string of the form
&lt;chromosome&gt;[:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;]]
(with square brackets indicating optional
parts of the string). If any optional
part of the string is omitted, intuitive
defaults will be chosen.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.is_forward">
<code class="sig-name descname">is_forward</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.is_forward" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if this region is on the forward strand
of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if on ‘+’ strand, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.is_reverse">
<code class="sig-name descname">is_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.is_reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if this region is on the reverse strand
of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if on ‘-‘ strand, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.overlap">
<code class="sig-name descname">overlap</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the overlap in base pairs between this
region and another region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> to find overlap for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>overlap as int in base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.overlaps">
<code class="sig-name descname">overlaps</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if this region overlaps with the specified region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object or string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.set_attribute">
<code class="sig-name descname">set_attribute</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">value</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.set_attribute" title="Permalink to this definition">¶</a></dt>
<dd><p>Safely set an attribute on the <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object.</p>
<p>This automatically decodes bytes objects into UTF-8 strings.
If you do not care about this, you can also use
region.&lt;attribute&gt; = &lt;value&gt; directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute to be set</p></li>
<li><p><strong>value</strong> – Value of the attribute to be set</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.strand_string">
<em class="property">property </em><code class="sig-name descname">strand_string</code><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.strand_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the ‘strand’ attribute as string.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>strand as str (‘+’, ‘-‘, or ‘.’)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.three_prime">
<em class="property">property </em><code class="sig-name descname">three_prime</code><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.three_prime" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the position of the 3’ end of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
on the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.compatibility.cooler.LazyCoolerRegion.to_string">
<code class="sig-name descname">to_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.LazyCoolerRegion.to_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> to its string representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="fanc.compatibility.cooler.to_cooler">
<code class="sig-prename descclassname">fanc.compatibility.cooler.</code><code class="sig-name descname">to_cooler</code><span class="sig-paren">(</span><em class="sig-param">hic</em>, <em class="sig-param">path</em>, <em class="sig-param">balance=True</em>, <em class="sig-param">multires=True</em>, <em class="sig-param">resolutions=None</em>, <em class="sig-param">n_zooms=10</em>, <em class="sig-param">threads=1</em>, <em class="sig-param">chunksize=100000</em>, <em class="sig-param">max_resolution=5000000</em>, <em class="sig-param">natural_order=True</em>, <em class="sig-param">chromosomes=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.compatibility.cooler.to_cooler" title="Permalink to this definition">¶</a></dt>
<dd><p>Export Hi-C data as Cooler file.</p>
<p>Only contacts that have not been
filtered are exported. <a class="reference external" href="https://github.com/mirnylab/cooler/">https://github.com/mirnylab/cooler/</a></p>
<p>Single resolution files:
If input Hi-C matrix is uncorrected, the uncorrected matrix is stored.
If it is corrected, the uncorrected matrix is stored along with bias vector.
Cooler always calculates corrected matrix on-the-fly from the uncorrected
matrix and the bias vector.</p>
<p>Multi-resolution files (default):</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hic</strong> – Hi-C file in any compatible (RegionMatrixContainer) format</p></li>
<li><p><strong>path</strong> – Output path for cooler file</p></li>
<li><p><strong>balance</strong> – Include bias vector in cooler output (single res) or perform
iterative correction (multi res)</p></li>
<li><p><strong>multires</strong> – Generate a multi-resolution cooler file</p></li>
<li><p><strong>resolutions</strong> – Resolutions in bp (int) for multi-resolution cooler output</p></li>
<li><p><strong>chunksize</strong> – Number of pixels processed at a time in cooler</p></li>
<li><p><strong>kwargs</strong> – Additional arguments passed to cooler.iterative_correction</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../architecture.html" class="btn btn-neutral float-right" title="FAN-C Python architecture modules" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="juicer.html" class="btn btn-neutral float-left" title="Juicer module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Kai Kruse, Vaquerizas lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>