

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Pairs module &mdash; FAN-C 0.9.10-beta documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Matrix module" href="matrix.html" />
    <link rel="prev" title="Regions module" href="regions.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> FAN-C
          

          
          </a>

          
            
            
              <div class="version">
                0.9.10
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting started with FAN-C</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fanc.html">fanc: command line tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fancplot.html">fancplot: plotting from the command line</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../compatibility.html">Working with Juicer and Cooler files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interfaces.html">Interfaces</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generate.html">Generate Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analyse.html">Analyse Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plot.html">Plotting (API)</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../modules.html">Reference</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="init.html">init</a></li>
<li class="toctree-l3"><a class="reference internal" href="map.html">map</a></li>
<li class="toctree-l3"><a class="reference internal" href="regions.html">regions</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">pairs</a></li>
<li class="toctree-l3"><a class="reference internal" href="matrix.html">matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="hic.html">hic</a></li>
<li class="toctree-l3"><a class="reference internal" href="peaks.html">peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="compatibility.html">compatibility</a></li>
<li class="toctree-l3"><a class="reference internal" href="architecture.html">architecture</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="tools.html">tools</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">FAN-C</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../api.html">fanc Python module</a> &raquo;</li>
        
          <li><a href="../modules.html">FAN-C Python modules</a> &raquo;</li>
        
      <li>Pairs module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/api/modules/pairs.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-fanc.pairs">
<span id="pairs-module"></span><h1>Pairs module<a class="headerlink" href="#module-fanc.pairs" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="fanc.pairs.BwaMemQualityFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">BwaMemQualityFilter</code><span class="sig-paren">(</span><em class="sig-param">cutoff=0.9</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.BwaMemQualityFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadFilter</span></code></a></p>
<p>Filters <code class="code docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">mem</span></code> generated alignments
based on the alignment score (normalized by
the length of the alignment).</p>
<dl class="method">
<dt id="fanc.pairs.BwaMemQualityFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.BwaMemQualityFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a read has a high alignment score.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.BwaMemUniquenessFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">BwaMemUniquenessFilter</code><span class="sig-paren">(</span><em class="sig-param">strict=False</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.BwaMemUniquenessFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadFilter</span></code></a></p>
<p>Filters <cite>bwa mem</cite> generated alignments based on whether they are unique or not.</p>
<dl class="method">
<dt id="fanc.pairs.BwaMemUniquenessFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.BwaMemUniquenessFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine if a read is valid or should be filtered.</p>
<p>When implementing custom read filters this method must be
overridden. It should return False for reads that
are to be filtered and True otherwise.</p>
<p>Internally, the ReadPairs object will iterate over all Read
instances to determine their validity on an individual
basis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">Read</span></code> object</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if Read is valid, False otherwise</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ContaminantFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">ContaminantFilter</code><span class="sig-paren">(</span><em class="sig-param">contaminant_reads</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ContaminantFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadFilter</span></code></a></p>
<p>Filter reads that also map to a contaminant genome.</p>
<dl class="method">
<dt id="fanc.pairs.ContaminantFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ContaminantFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a read also maps to a contaminant</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.FourDNucleomePairGenerator">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">FourDNucleomePairGenerator</code><span class="sig-paren">(</span><em class="sig-param">pairs_file</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FourDNucleomePairGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.TxtReadPairGenerator" title="fanc.pairs.TxtReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.TxtReadPairGenerator</span></code></a></p>
<p>Read pair generator that works on 4D Nucleome “.pairs” files.</p>
<p>For details on the 4D Nucleome pairs format see:
<a class="reference external" href="https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md">https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md</a></p>
<dl class="method">
<dt id="fanc.pairs.FourDNucleomePairGenerator.add_filter">
<code class="sig-name descname">add_filter</code><span class="sig-paren">(</span><em class="sig-param">read_filter</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FourDNucleomePairGenerator.add_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to this object.</p>
<p>The filter will be applied during read pair generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read_filter</strong> – <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FourDNucleomePairGenerator.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FourDNucleomePairGenerator.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Return filter statistics collected during read pair generation.</p>
<p>The <code class="xref py py-func docutils literal notranslate"><span class="pre">__iter__()</span></code> filters reads based on the
filters added using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>. During
filtering, it keeps track of the numbers of read pairs that were
filtered. This function returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> of the form
&lt;filter_name&gt;: &lt;number of reads filtered out&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.FragmentRead">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">FragmentRead</code><span class="sig-paren">(</span><em class="sig-param">fragment=None</em>, <em class="sig-param">position=None</em>, <em class="sig-param">strand=0</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentRead" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class representing a fragment-mapped read.</p>
<dl class="attribute">
<dt id="fanc.pairs.FragmentRead.fragment">
<code class="sig-name descname">fragment</code><a class="headerlink" href="#fanc.pairs.FragmentRead.fragment" title="Permalink to this definition">¶</a></dt>
<dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> delineated by
restriction sites.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.FragmentRead.position">
<code class="sig-name descname">position</code><a class="headerlink" href="#fanc.pairs.FragmentRead.position" title="Permalink to this definition">¶</a></dt>
<dd><p>The position of this read in base-pairs (1-based) from the
start of the chromosome it maps to.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.FragmentRead.strand">
<code class="sig-name descname">strand</code><a class="headerlink" href="#fanc.pairs.FragmentRead.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>The strand this read maps to (-1 or +1).</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FragmentRead.re_distance">
<code class="sig-name descname">re_distance</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentRead.re_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the distance of the alignment to the nearest restriction site.
:return: int</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.FragmentReadPair">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">FragmentReadPair</code><span class="sig-paren">(</span><em class="sig-param">left_read</em>, <em class="sig-param">right_read</em>, <em class="sig-param">ix=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Container for two paired <a class="reference internal" href="#fanc.pairs.FragmentRead" title="fanc.pairs.FragmentRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentRead</span></code></a> objects.</p>
<dl class="method">
<dt id="fanc.pairs.FragmentReadPair.get_gap_size">
<code class="sig-name descname">get_gap_size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair.get_gap_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the gap size in base pairs between the fragments these
reads map to.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>0 if reads map to the same fragment or neighboring
fragments, the distance between fragments if they
are on the same chromosome, None otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FragmentReadPair.is_inward_pair">
<code class="sig-name descname">is_inward_pair</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair.is_inward_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if reads form an “inward-facing” pair.</p>
<p>A pair is considered inward-facing if the left read maps
to the plus the right read to the minus strand and both
reads map to the same chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if reads are inward-facing, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FragmentReadPair.is_outward_pair">
<code class="sig-name descname">is_outward_pair</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair.is_outward_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if reads form an “outward-facing” pair.</p>
<p>A pair is considered outward-facing if the left read maps
to the minus the right read to the plus strand and both
reads map to the same chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if reads are outward-facing, False otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FragmentReadPair.is_same_chromosome">
<code class="sig-name descname">is_same_chromosome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair.is_same_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if both reads are mapped to the same chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True is reads map to the same chromosome, False
otherwise</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FragmentReadPair.is_same_fragment">
<code class="sig-name descname">is_same_fragment</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair.is_same_fragment" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if reads map to the same fragment.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if reads map to the same fragment,
False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.FragmentReadPair.is_same_pair">
<code class="sig-name descname">is_same_pair</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPair.is_same_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if reads face in the same direction.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if reads are facing in the same direction,
False otherwise.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.FragmentReadPairFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">FragmentReadPairFilter</code><span class="sig-paren">(</span><em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPairFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.general.MaskFilter</span></code></p>
<p>Abstract class that provides filtering functionality for the
<a class="reference internal" href="#fanc.pairs.FragmentReadPair" title="fanc.pairs.FragmentReadPair"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPair</span></code></a> object.</p>
<p>Extends MaskFilter and overrides valid(self, read) to make
<a class="reference internal" href="#fanc.pairs.FragmentReadPair" title="fanc.pairs.FragmentReadPair"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPair</span></code></a> filtering more “natural”.</p>
<p>To create custom filters for the <code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentMappedReadPairs</span></code>
object, extend this
class and override the <code class="xref py py-func docutils literal notranslate"><span class="pre">valid_pair()</span></code> method.
valid_pair should return False for a specific <a class="reference internal" href="#fanc.pairs.FragmentReadPair" title="fanc.pairs.FragmentReadPair"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPair</span></code></a> object
if the object is supposed to be filtered/masked and True
otherwise. See <a class="reference internal" href="#fanc.pairs.InwardPairsFilter" title="fanc.pairs.InwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">InwardPairsFilter</span></code></a> for an example.</p>
<p>Pass a custom filter to the filter method in <code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentMappedReadPairs</span></code>
to apply it.</p>
<dl class="method">
<dt id="fanc.pairs.FragmentReadPairFilter.valid">
<code class="sig-name descname">valid</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.FragmentReadPairFilter.valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Map validity check of rows to pairs.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.HicProPairGenerator">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">HicProPairGenerator</code><span class="sig-paren">(</span><em class="sig-param">file_name</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.HicProPairGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.TxtReadPairGenerator" title="fanc.pairs.TxtReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.TxtReadPairGenerator</span></code></a></p>
<p>Read pair generator for HiC-Pro “validPairs” files.</p>
<p>This generator is a subclass of <a class="reference internal" href="#fanc.pairs.TxtReadPairGenerator" title="fanc.pairs.TxtReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">TxtReadPairGenerator</span></code></a>
with presets for fields in HiC-Pro validPairs files.</p>
<dl class="method">
<dt id="fanc.pairs.HicProPairGenerator.add_filter">
<code class="sig-name descname">add_filter</code><span class="sig-paren">(</span><em class="sig-param">read_filter</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.HicProPairGenerator.add_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to this object.</p>
<p>The filter will be applied during read pair generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read_filter</strong> – <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.HicProPairGenerator.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.HicProPairGenerator.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Return filter statistics collected during read pair generation.</p>
<p>The <code class="xref py py-func docutils literal notranslate"><span class="pre">__iter__()</span></code> filters reads based on the
filters added using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>. During
filtering, it keeps track of the numbers of read pairs that were
filtered. This function returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> of the form
&lt;filter_name&gt;: &lt;number of reads filtered out&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.InwardPairsFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">InwardPairsFilter</code><span class="sig-paren">(</span><em class="sig-param">minimum_distance=10000</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.InwardPairsFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.FragmentReadPairFilter</span></code></a></p>
<p>Filter inward-facing read pairs at a distance less
than a specified cutoff.</p>
<dl class="method">
<dt id="fanc.pairs.InwardPairsFilter.valid">
<code class="sig-name descname">valid</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.InwardPairsFilter.valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Map validity check of rows to pairs.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.InwardPairsFilter.valid_pair">
<code class="sig-name descname">valid_pair</code><span class="sig-paren">(</span><em class="sig-param">pair</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.InwardPairsFilter.valid_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a pair is inward-facing and &lt;minimum_distance&gt; apart.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.LazyFragment">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">LazyFragment</code><span class="sig-paren">(</span><em class="sig-param">row</em>, <em class="sig-param">pairs</em>, <em class="sig-param">ix=None</em>, <em class="sig-param">side='left'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">genomic_regions.regions.GenomicRegion</span></code></p>
<p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> representing a fragment with lazy attribute loading.</p>
<dl class="attribute">
<dt id="fanc.pairs.LazyFragment.chromosome">
<code class="sig-name descname">chromosome</code><a class="headerlink" href="#fanc.pairs.LazyFragment.chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>The reference sequence this region is located on</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.LazyFragment.start">
<code class="sig-name descname">start</code><a class="headerlink" href="#fanc.pairs.LazyFragment.start" title="Permalink to this definition">¶</a></dt>
<dd><p>start position of this region on the reference (1-based, inclusive)</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.LazyFragment.end">
<code class="sig-name descname">end</code><a class="headerlink" href="#fanc.pairs.LazyFragment.end" title="Permalink to this definition">¶</a></dt>
<dd><p>end position of this region on the reference (1-based, inclusive)</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.LazyFragment.strand">
<code class="sig-name descname">strand</code><a class="headerlink" href="#fanc.pairs.LazyFragment.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>strand of the reference this region is located on (-1, 1, 0, or None)</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.LazyFragment.ix">
<code class="sig-name descname">ix</code><a class="headerlink" href="#fanc.pairs.LazyFragment.ix" title="Permalink to this definition">¶</a></dt>
<dd><p>Region index within the context of regions from the same object.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.as_bed_line">
<code class="sig-name descname">as_bed_line</code><span class="sig-paren">(</span><em class="sig-param">score_field='score'</em>, <em class="sig-param">name_field='name'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.as_bed_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of this object as line in a BED file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>score_field</strong> – name of the attribute to be used in the
‘score’ field of the BED line</p></li>
<li><p><strong>name_field</strong> – name of the attribute to be used in the
‘name’ field of the BED line</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.as_gff_line">
<code class="sig-name descname">as_gff_line</code><span class="sig-paren">(</span><em class="sig-param">source_field='source'</em>, <em class="sig-param">feature_field='feature'</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">frame_field='frame'</em>, <em class="sig-param">float_format='.2e'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.as_gff_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of this object as line in a GFF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source_field</strong> – name of the attribute to be used in the
‘source’ field of the GFF line</p></li>
<li><p><strong>feature_field</strong> – name of the attribute to be used in the
‘feature’ field of the GFF line</p></li>
<li><p><strong>score_field</strong> – name of the attribute to be used in the
‘score’ field of the GFF line</p></li>
<li><p><strong>frame_field</strong> – name of the attribute to be used in the
‘frame’ field of the GFF line</p></li>
<li><p><strong>float_format</strong> – Formatting string for the float fields</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.attributes">
<em class="property">property </em><code class="sig-name descname">attributes</code><a class="headerlink" href="#fanc.pairs.LazyFragment.attributes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all visible attributes of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<p>Returns all attribute names that do not start with an underscore.
:return: list of attribute names</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.center">
<em class="property">property </em><code class="sig-name descname">center</code><a class="headerlink" href="#fanc.pairs.LazyFragment.center" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the center coordinate of the <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the specified region is completely contained in this region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object or string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a (shallow) copy of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.expand">
<code class="sig-name descname">expand</code><span class="sig-paren">(</span><em class="sig-param">absolute=None</em>, <em class="sig-param">relative=None</em>, <em class="sig-param">absolute_left=0</em>, <em class="sig-param">absolute_right=0</em>, <em class="sig-param">relative_left=0.0</em>, <em class="sig-param">relative_right=0.0</em>, <em class="sig-param">copy=True</em>, <em class="sig-param">from_center=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.expand" title="Permalink to this definition">¶</a></dt>
<dd><p>Expand this region by a relative or an absolute amount.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>absolute</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on both
sides. New region start will be
&lt;old start - absolute&gt;, new region end
will be &lt;old end + absolute&gt;</p></li>
<li><p><strong>relative</strong> – Relative amount as fraction of region by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on both
sides. New region start will be
&lt;old start - relative*len(self)&gt;,
new region end will be
&lt;old end + relative*(len(self)&gt;</p></li>
<li><p><strong>absolute_left</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
left side</p></li>
<li><p><strong>absolute_right</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
right side</p></li>
<li><p><strong>relative_left</strong> – Relative amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
left side</p></li>
<li><p><strong>relative_right</strong> – Relative amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
right side</p></li>
<li><p><strong>copy</strong> – If True, return a copy of the original region,
if False will modify the existing region in place</p></li>
<li><p><strong>from_center</strong> – If True measures distance from center rather
than start and end of the old region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.five_prime">
<em class="property">property </em><code class="sig-name descname">five_prime</code><a class="headerlink" href="#fanc.pairs.LazyFragment.five_prime" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the position of the 5’ end of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
on the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.fix_chromosome">
<code class="sig-name descname">fix_chromosome</code><span class="sig-paren">(</span><em class="sig-param">copy=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.fix_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Change chromosome representation from chr&lt;NN&gt; to &lt;NN&gt; or vice versa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>copy</strong> – If True, make copy of region, otherwise
will modify existing region in place.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">region_string</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string into a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<p>This is a very useful convenience function to quickly
define a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object from a descriptor
string. Numbers can be abbreviated as ‘12k’, ‘1.5M’, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region_string</strong> – A string of the form
&lt;chromosome&gt;[:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;]]
(with square brackets indicating optional
parts of the string). If any optional
part of the string is omitted, intuitive
defaults will be chosen.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.is_forward">
<code class="sig-name descname">is_forward</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.is_forward" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if this region is on the forward strand
of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if on ‘+’ strand, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.is_reverse">
<code class="sig-name descname">is_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.is_reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if this region is on the reverse strand
of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if on ‘-‘ strand, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.overlap">
<code class="sig-name descname">overlap</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the overlap in base pairs between this
region and another region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> to find overlap for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>overlap as int in base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.overlaps">
<code class="sig-name descname">overlaps</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if this region overlaps with the specified region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object or string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.set_attribute">
<code class="sig-name descname">set_attribute</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">value</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.set_attribute" title="Permalink to this definition">¶</a></dt>
<dd><p>Safely set an attribute on the <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object.</p>
<p>This automatically decodes bytes objects into UTF-8 strings.
If you do not care about this, you can also use
region.&lt;attribute&gt; = &lt;value&gt; directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute to be set</p></li>
<li><p><strong>value</strong> – Value of the attribute to be set</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.strand_string">
<em class="property">property </em><code class="sig-name descname">strand_string</code><a class="headerlink" href="#fanc.pairs.LazyFragment.strand_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the ‘strand’ attribute as string.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>strand as str (‘+’, ‘-‘, or ‘.’)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.three_prime">
<em class="property">property </em><code class="sig-name descname">three_prime</code><a class="headerlink" href="#fanc.pairs.LazyFragment.three_prime" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the position of the 3’ end of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
on the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragment.to_string">
<code class="sig-name descname">to_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragment.to_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> to its string representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.LazyFragmentRead">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">LazyFragmentRead</code><span class="sig-paren">(</span><em class="sig-param">row</em>, <em class="sig-param">pairs</em>, <em class="sig-param">side='left'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragmentRead" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.FragmentRead" title="fanc.pairs.FragmentRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.FragmentRead</span></code></a></p>
<p><a class="reference internal" href="#fanc.pairs.FragmentRead" title="fanc.pairs.FragmentRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentRead</span></code></a> implementation with lazy attribute loading.</p>
<dl class="attribute">
<dt id="fanc.pairs.LazyFragmentRead.fragment">
<code class="sig-name descname">fragment</code><a class="headerlink" href="#fanc.pairs.LazyFragmentRead.fragment" title="Permalink to this definition">¶</a></dt>
<dd><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> delineated by
restriction sites.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.LazyFragmentRead.position">
<code class="sig-name descname">position</code><a class="headerlink" href="#fanc.pairs.LazyFragmentRead.position" title="Permalink to this definition">¶</a></dt>
<dd><p>The position of this read in base-pairs (1-based) from the
start of the chromosome it maps to.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.LazyFragmentRead.strand">
<code class="sig-name descname">strand</code><a class="headerlink" href="#fanc.pairs.LazyFragmentRead.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>The strand this read maps to (-1 or +1).</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.LazyFragmentRead.re_distance">
<code class="sig-name descname">re_distance</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.LazyFragmentRead.re_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the distance of the alignment to the nearest restriction site.
:return: int</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.MinimalRead">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">MinimalRead</code><span class="sig-paren">(</span><em class="sig-param">chromosome</em>, <em class="sig-param">position</em>, <em class="sig-param">strand</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.MinimalRead" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Minimal class representing an aligned read.</p>
<dl class="attribute">
<dt id="fanc.pairs.MinimalRead.chromosome">
<code class="sig-name descname">chromosome</code><a class="headerlink" href="#fanc.pairs.MinimalRead.chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Chromosome name</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.MinimalRead.reference_name">
<code class="sig-name descname">reference_name</code><a class="headerlink" href="#fanc.pairs.MinimalRead.reference_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Identical to chromosome, exists for
compatibility with pysam</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.MinimalRead.position">
<code class="sig-name descname">position</code><a class="headerlink" href="#fanc.pairs.MinimalRead.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Position of aligned read on chromosome</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.MinimalRead.pos">
<code class="sig-name descname">pos</code><a class="headerlink" href="#fanc.pairs.MinimalRead.pos" title="Permalink to this definition">¶</a></dt>
<dd><p>Identical to position, exists for
compatibility with Pysam</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.MinimalRead.strand">
<code class="sig-name descname">strand</code><a class="headerlink" href="#fanc.pairs.MinimalRead.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Strand of the alignment</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.pairs.MinimalRead.flag">
<code class="sig-name descname">flag</code><a class="headerlink" href="#fanc.pairs.MinimalRead.flag" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag representing the strandedness of a read.
0 if on forward strand, else -1</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.Monitor">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">Monitor</code><span class="sig-paren">(</span><em class="sig-param">value=0</em>, <em class="sig-param">manager=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.Monitor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.tools.general.WorkerMonitor</span></code></p>
<p>Class to monitor fragment info worker threads.</p>
<dl class="method">
<dt id="fanc.pairs.Monitor.is_generating_pairs">
<code class="sig-name descname">is_generating_pairs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.Monitor.is_generating_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Check the pair generating status.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.Monitor.set_generating_pairs">
<code class="sig-name descname">set_generating_pairs</code><span class="sig-paren">(</span><em class="sig-param">value</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.Monitor.set_generating_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the pair generating status.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.OutwardPairsFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">OutwardPairsFilter</code><span class="sig-paren">(</span><em class="sig-param">minimum_distance=10000</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.OutwardPairsFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.FragmentReadPairFilter</span></code></a></p>
<p>Filter outward-facing read pairs at a distance less
than a specified cutoff.</p>
<dl class="method">
<dt id="fanc.pairs.OutwardPairsFilter.valid">
<code class="sig-name descname">valid</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.OutwardPairsFilter.valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Map validity check of rows to pairs.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.PCRDuplicateFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">PCRDuplicateFilter</code><span class="sig-paren">(</span><em class="sig-param">pairs</em>, <em class="sig-param">threshold=2</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PCRDuplicateFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.FragmentReadPairFilter</span></code></a></p>
<p>Masks alignments that are suspected to be PCR duplicates.
In order to be considered duplicates, two pairs need to have identical
start positions of their respective left alignments AND of their right alignments.</p>
<dl class="method">
<dt id="fanc.pairs.PCRDuplicateFilter.valid">
<code class="sig-name descname">valid</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PCRDuplicateFilter.valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Map validity check of rows to pairs.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.PCRDuplicateFilter.valid_pair">
<code class="sig-name descname">valid_pair</code><span class="sig-paren">(</span><em class="sig-param">pair</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PCRDuplicateFilter.valid_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a pair is duplicated.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.PairedSamBamReadPairGenerator">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">PairedSamBamReadPairGenerator</code><span class="sig-paren">(</span><em class="sig-param">sam_file</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PairedSamBamReadPairGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadPairGenerator</span></code></a></p>
<p>Generate read pairs from single SAM/BAM files from reads with identical qname.</p>
<p>Chimeric reads (mapping partially to multiple genomic locations),
such as output by BWA, are handled as follows: If both mate pairs
are chimeric, they are removed. If only one mate is chimeric, but
it is split into more than 2 alignments, it is also removed. If
one mate is chimeric and split into two alignments. If one part
of the chimeric alignment maps within 100bp of the regular alignment,
the read pair is kept and returned. In all other cases, the pair is
removed.</p>
<dl class="method">
<dt id="fanc.pairs.PairedSamBamReadPairGenerator.add_filter">
<code class="sig-name descname">add_filter</code><span class="sig-paren">(</span><em class="sig-param">read_filter</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PairedSamBamReadPairGenerator.add_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to this object.</p>
<p>The filter will be applied during read pair generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read_filter</strong> – <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.PairedSamBamReadPairGenerator.resolve_chimeric">
<em class="property">static </em><code class="sig-name descname">resolve_chimeric</code><span class="sig-paren">(</span><em class="sig-param">reads</em>, <em class="sig-param">max_dist_same_locus=500</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PairedSamBamReadPairGenerator.resolve_chimeric" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>read1, read2, is_chimeric</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.PairedSamBamReadPairGenerator.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.PairedSamBamReadPairGenerator.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Return filter statistics collected during read pair generation.</p>
<p>The <code class="xref py py-func docutils literal notranslate"><span class="pre">__iter__()</span></code> filters reads based on the
filters added using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>. During
filtering, it keeps track of the numbers of read pairs that were
filtered. This function returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> of the form
&lt;filter_name&gt;: &lt;number of reads filtered out&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.QualityFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">QualityFilter</code><span class="sig-paren">(</span><em class="sig-param">cutoff=30</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.QualityFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadFilter</span></code></a></p>
<p>Filter mapped reads based on mapping quality.</p>
<dl class="method">
<dt id="fanc.pairs.QualityFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.QualityFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a read has a mapq &gt;= cutoff.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ReDistanceFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">ReDistanceFilter</code><span class="sig-paren">(</span><em class="sig-param">maximum_distance=10000</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReDistanceFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.FragmentReadPairFilter</span></code></a></p>
<p>Filters read pairs where the sum of distances of each read to its nearest
restriction site is larger than a certain distance.</p>
<dl class="method">
<dt id="fanc.pairs.ReDistanceFilter.valid">
<code class="sig-name descname">valid</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReDistanceFilter.valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Map validity check of rows to pairs.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReDistanceFilter.valid_pair">
<code class="sig-name descname">valid_pair</code><span class="sig-paren">(</span><em class="sig-param">pair</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReDistanceFilter.valid_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check sum of restriction site distances &lt; maximum_distance.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ReadFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">ReadFilter</code><span class="sig-paren">(</span><em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract class that provides filtering functionality for
<a class="reference internal" href="#fanc.pairs.MinimalRead" title="fanc.pairs.MinimalRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">MinimalRead</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">AlignedSegment</span></code> or
compatible.</p>
<p>To create a custom <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a>, extend this
class and override the valid_read(self, read) method.
valid_read should return False for a specific read object
if the object is supposed to be filtered/masked and True
otherwise. See <a class="reference internal" href="#fanc.pairs.QualityFilter" title="fanc.pairs.QualityFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">QualityFilter</span></code></a> for an example.</p>
<p>Pass a custom filter to the filter method in
<a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a> to apply it.</p>
<dl class="method">
<dt id="fanc.pairs.ReadFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine if a read is valid or should be filtered.</p>
<p>When implementing custom read filters this method must be
overridden. It should return False for reads that
are to be filtered and True otherwise.</p>
<p>Internally, the ReadPairs object will iterate over all Read
instances to determine their validity on an individual
basis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">Read</span></code> object</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if Read is valid, False otherwise</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ReadPairGenerator">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">ReadPairGenerator</code><a class="headerlink" href="#fanc.pairs.ReadPairGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for generating and filtering read pairs.</p>
<p>This class primarily provides filtering capabilities for
read pairs generated by subclasses of <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a>.
You can add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>,
which will be used during read pair generation. Filtering statistics
are collected during the run, and can be obtained via
<a class="reference internal" href="#fanc.pairs.ReadPairGenerator.stats" title="fanc.pairs.ReadPairGenerator.stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">stats()</span></code></a>.</p>
<p>These generators are primarily meant as input for
<a class="reference internal" href="#fanc.pairs.ReadPairs.add_read_pairs" title="fanc.pairs.ReadPairs.add_read_pairs"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_read_pairs()</span></code></a>, but can also be used
independently.</p>
<p>Subclasses of <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a> must implement the
<code class="xref py py-func docutils literal notranslate"><span class="pre">_iter_read_pairs()</span></code> function.</p>
<dl class="method">
<dt id="fanc.pairs.ReadPairGenerator.add_filter">
<code class="sig-name descname">add_filter</code><span class="sig-paren">(</span><em class="sig-param">read_filter</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairGenerator.add_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to this object.</p>
<p>The filter will be applied during read pair generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read_filter</strong> – <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairGenerator.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairGenerator.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Return filter statistics collected during read pair generation.</p>
<p>The <code class="xref py py-func docutils literal notranslate"><span class="pre">__iter__()</span></code> filters reads based on the
filters added using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>. During
filtering, it keeps track of the numbers of read pairs that were
filtered. This function returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> of the form
&lt;filter_name&gt;: &lt;number of reads filtered out&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ReadPairs">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">ReadPairs</code><span class="sig-paren">(</span><em class="sig-param">file_name=None</em>, <em class="sig-param">mode='a'</em>, <em class="sig-param">partition_strategy='auto'</em>, <em class="sig-param">_group_name='fragment_map'</em>, <em class="sig-param">_table_name_fragments='fragments'</em>, <em class="sig-param">_table_name_pairs='pairs'</em>, <em class="sig-param">tmpdir=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="matrix.html#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.RegionPairsTable</span></code></a></p>
<p>Class representing a collection of read pairs mapped to restriction fragments.</p>
<p>This class is a <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code> object, where each
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> represents a restriction fragment
from a Hi-C experiment. A list of fragments can be obtained with
the <a class="reference internal" href="regions.html#fanc.regions.genome_regions" title="fanc.regions.genome_regions"><code class="xref py py-func docutils literal notranslate"><span class="pre">genome_regions()</span></code></a> function, for example.</p>
<p>To create a <a class="reference internal" href="#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a> object, you first have to add
the restriction fragments before adding read pairs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="n">re_fragments</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">genome_regions</span><span class="p">(</span><span class="s2">&quot;hg19_chr18_19.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;HindIII&quot;</span><span class="p">)</span>

<span class="n">rp</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">ReadPairs</span><span class="p">()</span>
<span class="n">rp</span><span class="o">.</span><span class="n">add_regions</span><span class="p">(</span><span class="n">re_fragments</span><span class="o">.</span><span class="n">regions</span><span class="p">)</span>
</pre></div>
</div>
<p>Read pairs can easily be generate fro different types of input using
<a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a> implementations, e.g.
<code class="xref py py-class docutils literal notranslate"><span class="pre">HicProReadPairGenerator</span></code> or <a class="reference internal" href="#fanc.pairs.SamBamReadPairGenerator" title="fanc.pairs.SamBamReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">SamBamReadPairGenerator</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rp_generator</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">SamBamReadPairGenerator</span><span class="p">(</span><span class="s2">&quot;output/sam/SRR4271982_chr18_19_1_sort.bam&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;output/sam/SRR4271982_chr18_19_2_sort.bam&quot;</span><span class="p">)</span>
<span class="n">rp</span><span class="o">.</span><span class="n">add_read_pairs</span><span class="p">(</span><span class="n">rp_generator</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<p>You can query regions using the <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code> interface:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr1_fragments</span> <span class="o">=</span> <span class="n">rp</span><span class="o">.</span><span class="n">regions</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>and you can iterate over read pairs using the <a class="reference internal" href="#fanc.pairs.ReadPairs.pairs" title="fanc.pairs.ReadPairs.pairs"><code class="xref py py-func docutils literal notranslate"><span class="pre">pairs()</span></code></a>:</p>
<p>you can also use the :cls:Region</p>
<dl class="class">
<dt id="fanc.pairs.ReadPairs.ChromosomeDescription">
<em class="property">class </em><code class="sig-name descname">ChromosomeDescription</code><a class="headerlink" href="#fanc.pairs.ReadPairs.ChromosomeDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ReadPairs.MaskDescription">
<em class="property">class </em><code class="sig-name descname">MaskDescription</code><a class="headerlink" href="#fanc.pairs.ReadPairs.MaskDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
</dd></dl>

<dl class="class">
<dt id="fanc.pairs.ReadPairs.RegionDescription">
<em class="property">class </em><code class="sig-name descname">RegionDescription</code><a class="headerlink" href="#fanc.pairs.ReadPairs.RegionDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_contact">
<code class="sig-name descname">add_contact</code><span class="sig-paren">(</span><em class="sig-param">contact</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_contact" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contact</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_contacts">
<code class="sig-name descname">add_contacts</code><span class="sig-paren">(</span><em class="sig-param">contacts</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <code class="xref py py-func docutils literal notranslate"><span class="pre">add_edges()</span></code></p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_edge">
<code class="sig-name descname">add_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">check_nodes_exist=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an edge / contact between two regions to this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code>, dict with at least the
attributes source and sink, optionally weight,
or a list of length 2 (source, sink) or 3
(source, sink, weight).</p></li>
<li><p><strong>check_nodes_exist</strong> – Make sure that there are nodes
that match source and sink indexes</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_edge_from_dict">
<code class="sig-name descname">add_edge_from_dict</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_edge_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from dict input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – dict with at least the keys “source”
and “sink”. Additional keys will be loaded
as edge attributes</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_edge_from_edge">
<code class="sig-name descname">add_edge_from_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_edge_from_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_edge_from_list">
<code class="sig-name descname">add_edge_from_list</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_edge_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from list or tuple input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – List or tuple. Should be of length 2
(source, sink) or 3 (source, sink, weight)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_edge_simple">
<code class="sig-name descname">add_edge_simple</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">weight=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_edge_simple" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> – Source region index</p></li>
<li><p><strong>sink</strong> – Sink region index</p></li>
<li><p><strong>weight</strong> – Weight of the edge</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_edges">
<code class="sig-name descname">add_edges</code><span class="sig-paren">(</span><em class="sig-param">edges</em>, <em class="sig-param">flush=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk-add edges from a list.</p>
<p>List items can be any of the supported edge types,
list, tuple, dict, or <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code>. Repeatedly
calls <code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code>, so
may be inefficient for large amounts of data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edges</strong> – List (or iterator) of edges. See
<code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code>
for details</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_mask_description">
<code class="sig-name descname">add_mask_description</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">description</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_mask_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a mask description to the _mask table and return its ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of the mask</p></li>
<li><p><strong>description</strong> (<em>str</em>) – description of the mask</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>id of the mask</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_read_pairs">
<code class="sig-name descname">add_read_pairs</code><span class="sig-paren">(</span><em class="sig-param">read_pairs</em>, <em class="sig-param">batch_size=1000000</em>, <em class="sig-param">threads=1</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_read_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Add read pairs to this object.</p>
<p>This function requires tuples of read pairs as input, for
example from <a class="reference internal" href="#fanc.pairs.MinimalRead" title="fanc.pairs.MinimalRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">MinimalRead</span></code></a> or <code class="xref py py-class docutils literal notranslate"><span class="pre">AlignedSegment</span></code>.
Typically, you won’t have to construct these from scratch, but can
use one of the <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a> classes to generate
read pairs from input files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="n">re_fragments</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">genome_regions</span><span class="p">(</span><span class="s2">&quot;hg19_chr18_19.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;HindIII&quot;</span><span class="p">)</span>
<span class="n">rp</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">ReadPairs</span><span class="p">()</span>
<span class="n">rp</span><span class="o">.</span><span class="n">add_regions</span><span class="p">(</span><span class="n">re_fragments</span><span class="o">.</span><span class="n">regions</span><span class="p">)</span>

<span class="c1"># read pairs are added here from BAM files</span>
<span class="n">rp_generator</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">SamBamReadPairGenerator</span><span class="p">(</span><span class="s2">&quot;output/sam/SRR4271982_chr18_19_1_sort.bam&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;output/sam/SRR4271982_chr18_19_2_sort.bam&quot;</span><span class="p">)</span>
<span class="n">rp</span><span class="o">.</span><span class="n">add_read_pairs</span><span class="p">(</span><span class="n">rp_generator</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_pairs</strong> – iterator over tuples of read pairs. Typically
instances of <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a></p></li>
<li><p><strong>batch_size</strong> – Batch size of read pairs sent to fragment info workers</p></li>
<li><p><strong>threads</strong> – Number of threads for simultaneous fragment info finding</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.add_regions">
<code class="sig-name descname">add_regions</code><span class="sig-paren">(</span><em class="sig-param">regions</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.add_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.bin_size">
<em class="property">property </em><code class="sig-name descname">bin_size</code><a class="headerlink" href="#fanc.pairs.ReadPairs.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.bins_to_distance">
<code class="sig-name descname">bins_to_distance</code><span class="sig-paren">(</span><em class="sig-param">bins</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.bins_to_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.chromosome_bins">
<em class="property">property </em><code class="sig-name descname">chromosome_bins</code><a class="headerlink" href="#fanc.pairs.ReadPairs.chromosome_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.pairs.ReadPairs.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><em class="sig-param">copy_tmp=True</em>, <em class="sig-param">remove_tmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.distance_to_bins">
<code class="sig-name descname">distance_to_bins</code><span class="sig-paren">(</span><em class="sig-param">distance</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.distance_to_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.downsample">
<code class="sig-name descname">downsample</code><span class="sig-paren">(</span><em class="sig-param">n</em>, <em class="sig-param">file_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.downsample" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample edges from this object.</p>
<p>Sampling is always done on uncorrected Hi-C matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> – Sample size or reference object. If n &lt; 1 will be interpreted as
a fraction of total reads in this object.</p></li>
<li><p><strong>file_name</strong> – Output file name for down-sampled object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.edge_data">
<code class="sig-name descname">edge_data</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.edge_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over specific edge attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute, e.g. “weight”</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator over edge attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.edge_subset">
<code class="sig-name descname">edge_subset</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.edge_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subset of edges.</p>
<p>This is an alias for <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generator (<code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.edges">
<em class="property">property </em><code class="sig-name descname">edges</code><a class="headerlink" href="#fanc.pairs.ReadPairs.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over contacts / edges.</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code> is the central function of
<code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code>. Here, we will use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> implementation for demonstration purposes,
but the usage is exactly the same for all compatible
objects implementing <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code>, including
<a class="reference internal" href="compatibility/juicer.html#fanc.compatibility.juicer.JuicerHic" title="fanc.compatibility.juicer.JuicerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">JuicerHic</span></code></a> and
<a class="reference internal" href="compatibility/cooler.html#fanc.compatibility.cooler.CoolerHic" title="fanc.compatibility.cooler.CoolerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">CoolerHic</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># file from FAN-C examples</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can easily find the number of edges in the sample
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span>  <span class="c1"># 8695</span>
</pre></div>
</div>
<p>When used in an iterator context, <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>
iterates over all edges in the <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Calling <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code> as a method has the
same effect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># note the &#39;()&#39;</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Rather than iterate over all edges in the object, we can select only a subset.
If the key is a string or a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, all non-zero edges connecting
the region described by the key to any other region are returned. If the key is a
tuple of strings or <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, only edges between the two regions
are returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all edges between chromosome 19</span>
<span class="c1"># and any other region:</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>

<span class="c1"># select all edges that are only on</span>
<span class="c1"># chromosome 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 90--116; bias: 0.00021173151730025176; sink_node: chr19:37000001-38000000; source_node: chr19:11000001-12000000; weight: 0.009104455243910825</span>
    <span class="c1"># 135--135; bias: 0.00018003890596887822; sink_node: chr19:56000001-57000000; source_node: chr19:56000001-57000000; weight: 0.10028167062466517</span>
    <span class="c1"># 123--123; bias: 0.00011063368998965993; sink_node: chr19:44000001-45000000; source_node: chr19:44000001-45000000; weight: 0.1386240135570439</span>
    <span class="c1"># 92--93; bias: 0.00040851066434864896; sink_node: chr19:14000001-15000000; source_node: chr19:13000001-14000000; weight: 0.10090213409411629</span>
    <span class="c1"># ...</span>

<span class="c1"># select inter-chromosomal edges</span>
<span class="c1"># between chromosomes 18 and 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>By default, <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code> will retrieve all edge attributes,
which can be slow when iterating over a lot of edges. This is why all file-based FAN-C
<code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code> objects support lazy loading, where attributes
are only read on demand.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span> <span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42 42 0.12291311562018173 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #0&gt;</span>
    <span class="c1"># 24 28 0.025205961072838057 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #1&gt;</span>
    <span class="c1"># 5 76 0.00961709840049876 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #2&gt;</span>
    <span class="c1"># 66 68 0.03876763316345468 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #3&gt;</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The lazy iterator reuses the <code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code> object in every iteration,
and overwrites the <code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code> attributes. Therefore <strong>do not</strong> use
lazy iterators if you need to store edge objects for later access.
For example, the following code works as expected
<code class="code docutils literal notranslate"><span class="pre">list(hic.edges())</span></code>, with all <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> objects stored in the
list, while this code <code class="code docutils literal notranslate"><span class="pre">list(hic.edges(lazy=True))</span></code>
will result in a list of identical <code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code> objects. Always ensure
you do all edge processing in the loop when working with lazy iterators!</p>
</div>
<p>When working with normalised contact frequencies, such as obtained through
matrix balancing in the example above, <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code>
automatically returns normalised edge weights. In addition, the <code class="code docutils literal notranslate"><span class="pre">bias</span></code>
attribute will (typically) have a value different from 1.</p>
<p>When you are interested in the raw contact frequency, use the <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>
parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">)</span>
    <span class="c1"># 42 42 2120.0</span>
    <span class="c1"># 24 28 388.0</span>
    <span class="c1"># 5 76 94.0</span>
    <span class="c1"># 66 68 470.0</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>You can also choose to omit all intra- or inter-chromosomal edges using
<code class="code docutils literal notranslate"><span class="pre">intra_chromosomal=False</span></code> or <code class="code docutils literal notranslate"><span class="pre">inter_chromosomal=False</span></code>, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Iterator over <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> or equivalent.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.edges_dict">
<code class="sig-name descname">edges_dict</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.edges_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Edges iterator with access by bracket notation.</p>
<p>This iterator <strong>always</strong> returns unnormalised edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict or dict-like iterator</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter">
<code class="sig-name descname">filter</code><span class="sig-paren">(</span><em class="sig-param">pair_filter</em>, <em class="sig-param">queue=False</em>, <em class="sig-param">log_progress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply a <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPairFilter</span></code></a> to the read pairs in this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pair_filter</strong> – <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPairFilter</span></code></a></p></li>
<li><p><strong>queue</strong> – If True, does not do the filtering immediately, but
queues this filter. All queued filters can then be run
at the same time using <a class="reference internal" href="#fanc.pairs.ReadPairs.run_queued_filters" title="fanc.pairs.ReadPairs.run_queued_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_queued_filters()</span></code></a></p></li>
<li><p><strong>log_progress</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_inward">
<code class="sig-name descname">filter_inward</code><span class="sig-paren">(</span><em class="sig-param">minimum_distance=None</em>, <em class="sig-param">queue=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_inward" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that applies an <a class="reference internal" href="#fanc.pairs.InwardPairsFilter" title="fanc.pairs.InwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">InwardPairsFilter</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>minimum_distance</strong> – Minimum distance inward-facing read
pairs must have to pass this filter</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
run_queued_filters</p></li>
<li><p><strong>kwargs</strong> – Additional arguments to pass
to <a class="reference internal" href="#fanc.pairs.ReadPairs.get_ligation_structure_biases" title="fanc.pairs.ReadPairs.get_ligation_structure_biases"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_ligation_structure_biases()</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_ligation_products">
<code class="sig-name descname">filter_ligation_products</code><span class="sig-paren">(</span><em class="sig-param">inward_threshold=None</em>, <em class="sig-param">outward_threshold=None</em>, <em class="sig-param">queue=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_ligation_products" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that applies an <a class="reference internal" href="#fanc.pairs.OutwardPairsFilter" title="fanc.pairs.OutwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">OutwardPairsFilter</span></code></a> and an <a class="reference internal" href="#fanc.pairs.InwardPairsFilter" title="fanc.pairs.InwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">InwardPairsFilter</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>inward_threshold</strong> – Minimum distance inward-facing read
pairs must have to pass this filter.
If None, will be inferred from the data</p></li>
<li><p><strong>outward_threshold</strong> – Minimum distance outward-facing read
pairs must have to pass this filter.
If None, will be inferred from the data</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
run_queued_filters</p></li>
<li><p><strong>kwargs</strong> – Additional arguments to pass
to <a class="reference internal" href="#fanc.pairs.ReadPairs.get_ligation_structure_biases" title="fanc.pairs.ReadPairs.get_ligation_structure_biases"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_ligation_structure_biases()</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_outward">
<code class="sig-name descname">filter_outward</code><span class="sig-paren">(</span><em class="sig-param">minimum_distance=None</em>, <em class="sig-param">queue=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_outward" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that applies an <a class="reference internal" href="#fanc.pairs.OutwardPairsFilter" title="fanc.pairs.OutwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">OutwardPairsFilter</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>minimum_distance</strong> – Minimum distance outward-facing read
pairs must have to pass this filter</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
run_queued_filters</p></li>
<li><p><strong>kwargs</strong> – Additional arguments to pass
to <a class="reference internal" href="#fanc.pairs.ReadPairs.get_ligation_structure_biases" title="fanc.pairs.ReadPairs.get_ligation_structure_biases"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_ligation_structure_biases()</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_pcr_duplicates">
<code class="sig-name descname">filter_pcr_duplicates</code><span class="sig-paren">(</span><em class="sig-param">threshold=3</em>, <em class="sig-param">queue=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_pcr_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that applies an <a class="reference internal" href="#fanc.pairs.PCRDuplicateFilter" title="fanc.pairs.PCRDuplicateFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">PCRDuplicateFilter</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>threshold</strong> – If distance between two alignments is smaller or
equal the threshold, the alignments
are considered to be starting at the same position</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
run_queued_filters</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_re_dist">
<code class="sig-name descname">filter_re_dist</code><span class="sig-paren">(</span><em class="sig-param">maximum_distance</em>, <em class="sig-param">queue=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_re_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that applies an <a class="reference internal" href="#fanc.pairs.ReDistanceFilter" title="fanc.pairs.ReDistanceFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReDistanceFilter</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>maximum_distance</strong> – Maximum distance a read can have to the
nearest region border (=restriction site)</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
run_queued_filters</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_self_ligated">
<code class="sig-name descname">filter_self_ligated</code><span class="sig-paren">(</span><em class="sig-param">queue=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_self_ligated" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that applies an <a class="reference internal" href="#fanc.pairs.SelfLigationFilter" title="fanc.pairs.SelfLigationFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">SelfLigationFilter</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
run_queued_filters</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.filter_statistics">
<code class="sig-name descname">filter_statistics</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.filter_statistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Get filtering statistics for this object.
:return: dict with {filter_type: count, …}</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.flush">
<code class="sig-name descname">flush</code><span class="sig-paren">(</span><em class="sig-param">silent=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.flush" title="Permalink to this definition">¶</a></dt>
<dd><p>Write buffered data to file and update indexes,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>silent</strong> – If True, does not use progressbars.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.get_edge">
<code class="sig-name descname">get_edge</code><span class="sig-paren">(</span><em class="sig-param">item</em>, <em class="sig-param">*row_conversion_args</em>, <em class="sig-param">**row_conversion_kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.get_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an edge by index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>row_conversion_args</strong> – Arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">RegionPairs._row_to_edge()</span></code></p></li>
<li><p><strong>row_conversion_kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">RegionPairs._row_to_edge()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.get_ligation_structure_biases">
<code class="sig-name descname">get_ligation_structure_biases</code><span class="sig-paren">(</span><em class="sig-param">sampling=None</em>, <em class="sig-param">skip_self_ligations=True</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.get_ligation_structure_biases" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the ligation biases (inward and outward to same-strand) of this data set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sampling</strong> – Approximate number of data points to average per point
in the plot. If None (default), this will
be determined on a best-guess basis.</p></li>
<li><p><strong>skip_self_ligations</strong> – If True (default), will not consider
self-ligated fragments for assessing
the error rates.</p></li>
<li><p><strong>unfiltered</strong> – If True, uses all read pairs, even those that do not
pass filters, for the ligation error calculation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>tuple with (list of gap sizes between reads, list of matching le type ratios)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Search _mask table for key and return Mask.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<em>int</em>) – search by mask name</p></li>
<li><p><strong>key</strong> – search by mask ID</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Mask</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.get_masks">
<code class="sig-name descname">get_masks</code><span class="sig-paren">(</span><em class="sig-param">ix</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.get_masks" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract mask IDs encoded in parameter and return masks.</p>
<p>IDs are powers of 2, so a single int field in the table can hold
multiple masks by simply adding up the IDs. Similar principle to
UNIX chmod (although that uses base 8)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ix</strong> (<em>int</em>) – integer that is the sum of powers of 2. Note that this value
is not necessarily itself a power of 2.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of Masks extracted from ix</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list (Mask)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.mappable">
<code class="sig-name descname">mappable</code><span class="sig-paren">(</span><em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.mappable" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mappability of regions in this object.</p>
<p>A “mappable” region has at least one contact to another region
in the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code> where True means mappable
and False unmappable</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">pairs</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge two or more <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – list of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>merged <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.pairs">
<code class="sig-name descname">pairs</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">lazy=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over the <a class="reference internal" href="#fanc.pairs.FragmentReadPair" title="fanc.pairs.FragmentReadPair"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPair</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Region string of the form &lt;chromosome&gt;[:&lt;start&gt;-&lt;end&gt;],
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or tuples thereof</p></li>
<li><p><strong>lazy</strong> – If True, use lazy loading of objects and their attributes.
Much faster, but can lead to unexpected results if one is
not careful. For example, this: <code class="code docutils literal notranslate"><span class="pre">list(object.pairs())</span></code>
is not the same as <code class="code docutils literal notranslate"><span class="pre">list(object.pairs(lazy=True))</span></code>! In
the latter case, all objects in the list will be identical
due to the lazy loading process. Only use lazy loading to access
attributes in an iterator!</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges_dict()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges_dict()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.pairs_by_chromosomes">
<code class="sig-name descname">pairs_by_chromosomes</code><span class="sig-paren">(</span><em class="sig-param">chromosome1</em>, <em class="sig-param">chromosome2</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.pairs_by_chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Only iterate over read pairs in this combination of chromosomes.
:param chromosome1: Name of first chromosome
:param chromosome2: Name of second chromosome
:param kwargs: Keyword arguments passed to <a class="reference internal" href="#fanc.pairs.ReadPairs.pairs" title="fanc.pairs.ReadPairs.pairs"><code class="xref py py-func docutils literal notranslate"><span class="pre">pairs()</span></code></a>
:return:</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.region_data">
<code class="sig-name descname">region_data</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.region_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.pairs.ReadPairs.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.regions_and_edges">
<code class="sig-name descname">regions_and_edges</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.regions_and_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to regions and edges selected by key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge selector, see <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>args</strong> – Positional arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over edges</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.pairs.ReadPairs.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.regions_identical">
<em class="property">static </em><code class="sig-name descname">regions_identical</code><span class="sig-paren">(</span><em class="sig-param">pairs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.regions_identical" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the regions in all objects in the list are identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
objects</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if chromosome, start, and end are identical between
all regions in the same list positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.run_queued_filters">
<code class="sig-name descname">run_queued_filters</code><span class="sig-paren">(</span><em class="sig-param">log_progress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.run_queued_filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Run queued filters. See <a class="reference internal" href="#fanc.pairs.ReadPairs.filter" title="fanc.pairs.ReadPairs.filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>log_progress</strong> – If true, process iterating through all edges
will be continuously reported.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><em class="sig-param">*regions</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Subset a Hic object by specifying one or more subset regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>regions</strong> – string or GenomicRegion object(s)</p></li>
<li><p><strong>kwargs</strong> – Supports
file_name: destination file name of subset Hic object;
tmpdir: if True works in tmp until object is closed
additional parameters are passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.ReadPairs.to_hic">
<code class="sig-name descname">to_hic</code><span class="sig-paren">(</span><em class="sig-param">file_name=None</em>, <em class="sig-param">tmpdir=None</em>, <em class="sig-param">_hic_class=&lt;class 'fanc.hic.Hic'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.ReadPairs.to_hic" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this <a class="reference internal" href="#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a> to a <code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to the <code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> output file</p></li>
<li><p><strong>tmpdir</strong> – If True (or path to temporary directory) will
work in temporary directory until closed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.SamBamReadPairGenerator">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">SamBamReadPairGenerator</code><span class="sig-paren">(</span><em class="sig-param">sam_file1</em>, <em class="sig-param">sam_file2</em>, <em class="sig-param">check_sorted=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SamBamReadPairGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadPairGenerator</span></code></a></p>
<p>Generate read pairs from paired-end SAM/BAM files.</p>
<p>This <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a> iterates over two SAM or BAM
files that have been sorted by qname (for example with
<code class="code docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">sort</span> <span class="pre">-n</span></code>).</p>
<p>Chimeric reads (mapping partially to multiple genomic locations),
such as output by BWA, are handled as follows: If both mate pairs
are chimeric, they are removed. If only one mate is chimeric, but
it is split into more than 2 alignments, it is also removed. If
one mate is chimeric and split into two alignments. If one part
of the chimeric alignment maps within 100bp of the regular alignment,
the read pair is kept and returned. In all other cases, the pair is
removed.</p>
<dl class="method">
<dt id="fanc.pairs.SamBamReadPairGenerator.add_filter">
<code class="sig-name descname">add_filter</code><span class="sig-paren">(</span><em class="sig-param">read_filter</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SamBamReadPairGenerator.add_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to this object.</p>
<p>The filter will be applied during read pair generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read_filter</strong> – <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.SamBamReadPairGenerator.filter_multi_mapping">
<code class="sig-name descname">filter_multi_mapping</code><span class="sig-paren">(</span><em class="sig-param">strict=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SamBamReadPairGenerator.filter_multi_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that registers a UniquenessFilter.
The actual algorithm and rationale used for filtering will depend on the
internal _mapper attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>strict</strong> – If True will filter if XS tag is present. If False,
will filter only when XS tag is not 0. This is applied if
alignments are from bowtie2.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.SamBamReadPairGenerator.filter_quality">
<code class="sig-name descname">filter_quality</code><span class="sig-paren">(</span><em class="sig-param">cutoff=30</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SamBamReadPairGenerator.filter_quality" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that registers a QualityFilter.
The actual algorithm and rationale used for filtering will depend on the
internal _mapper attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cutoff</strong> – Minimum mapping quality (mapq) a read must have to pass
the filter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.SamBamReadPairGenerator.filter_unmapped">
<code class="sig-name descname">filter_unmapped</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SamBamReadPairGenerator.filter_unmapped" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function that registers an UnmappedFilter.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.SamBamReadPairGenerator.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SamBamReadPairGenerator.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Return filter statistics collected during read pair generation.</p>
<p>The <code class="xref py py-func docutils literal notranslate"><span class="pre">__iter__()</span></code> filters reads based on the
filters added using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>. During
filtering, it keeps track of the numbers of read pairs that were
filtered. This function returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> of the form
&lt;filter_name&gt;: &lt;number of reads filtered out&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.SelfLigationFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">SelfLigationFilter</code><span class="sig-paren">(</span><em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SelfLigationFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.FragmentReadPairFilter</span></code></a></p>
<p>Filters read pairs where one or both reads map to the
same restriction fragment.</p>
<dl class="method">
<dt id="fanc.pairs.SelfLigationFilter.valid">
<code class="sig-name descname">valid</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SelfLigationFilter.valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Map validity check of rows to pairs.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.SelfLigationFilter.valid_pair">
<code class="sig-name descname">valid_pair</code><span class="sig-paren">(</span><em class="sig-param">pair</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.SelfLigationFilter.valid_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if both reads are on the same fragment.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.TxtReadPairGenerator">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">TxtReadPairGenerator</code><span class="sig-paren">(</span><em class="sig-param">valid_pairs_file</em>, <em class="sig-param">sep=None</em>, <em class="sig-param">chr1_field=1</em>, <em class="sig-param">pos1_field=2</em>, <em class="sig-param">strand1_field=3</em>, <em class="sig-param">chr2_field=4</em>, <em class="sig-param">pos2_field=5</em>, <em class="sig-param">strand2_field=6</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.TxtReadPairGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadPairGenerator</span></code></a></p>
<p>Generate read pairs from a plain text file.</p>
<p>This is an implementation of <a class="reference internal" href="#fanc.pairs.ReadPairGenerator" title="fanc.pairs.ReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairGenerator</span></code></a> for
reading read pairs from an arbitrary text file. For specific
text file formats have a look at <a class="reference internal" href="#fanc.pairs.FourDNucleomePairGenerator" title="fanc.pairs.FourDNucleomePairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">FourDNucleomePairGenerator</span></code></a>
or <a class="reference internal" href="#fanc.pairs.HicProPairGenerator" title="fanc.pairs.HicProPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">HicProPairGenerator</span></code></a>.</p>
<p><a class="reference internal" href="#fanc.pairs.TxtReadPairGenerator" title="fanc.pairs.TxtReadPairGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">TxtReadPairGenerator</span></code></a> iterates over lines in
a file and splits them into fields at “sep”. It then extracts
the chromosome, position and strand for each read in the pair,
according to the fields specified by “chr&lt;1|2&gt;_field”,
“pos&lt;1|2&gt;_field”, and “strand&lt;1|2&gt;_field”. If your file does not
have strand fields, or if you don’t want to load them, you can
simply set them to “None”.</p>
<dl class="method">
<dt id="fanc.pairs.TxtReadPairGenerator.add_filter">
<code class="sig-name descname">add_filter</code><span class="sig-paren">(</span><em class="sig-param">read_filter</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.TxtReadPairGenerator.add_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to this object.</p>
<p>The filter will be applied during read pair generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read_filter</strong> – <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.pairs.TxtReadPairGenerator.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.TxtReadPairGenerator.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Return filter statistics collected during read pair generation.</p>
<p>The <code class="xref py py-func docutils literal notranslate"><span class="pre">__iter__()</span></code> filters reads based on the
filters added using <a class="reference internal" href="#fanc.pairs.ReadPairGenerator.add_filter" title="fanc.pairs.ReadPairGenerator.add_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_filter()</span></code></a>. During
filtering, it keeps track of the numbers of read pairs that were
filtered. This function returns a <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> of the form
&lt;filter_name&gt;: &lt;number of reads filtered out&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.UniquenessFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">UniquenessFilter</code><span class="sig-paren">(</span><em class="sig-param">strict=True</em>, <em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.UniquenessFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadFilter</span></code></a></p>
<p>Filter reads that do not map uniquely to the reference sequence.</p>
<dl class="method">
<dt id="fanc.pairs.UniquenessFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.UniquenessFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a read has an XS tag.</p>
<p>If strict is enabled checks if a read has an XS tag.
If not strict, XS has to be smaller than AS (alignment score)
for a valid read.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.pairs.UnmappedFilter">
<em class="property">class </em><code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">UnmappedFilter</code><span class="sig-paren">(</span><em class="sig-param">mask=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.UnmappedFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.pairs.ReadFilter</span></code></a></p>
<p>Filter reads that do not map to the reference sequence.</p>
<dl class="method">
<dt id="fanc.pairs.UnmappedFilter.valid_read">
<code class="sig-name descname">valid_read</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.UnmappedFilter.valid_read" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the the flag bit 4 is set.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="fanc.pairs.generate_pairs">
<code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">generate_pairs</code><span class="sig-paren">(</span><em class="sig-param">sam1_file</em>, <em class="sig-param">sam2_file</em>, <em class="sig-param">regions</em>, <em class="sig-param">restriction_enzyme=None</em>, <em class="sig-param">output_file=None</em>, <em class="sig-param">read_filters=()</em>, <em class="sig-param">check_sorted=True</em>, <em class="sig-param">threads=1</em>, <em class="sig-param">batch_size=10000000</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.generate_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate Pairs object from SAM/BAM files.</p>
<p>This is a convenience function that let’s you create a
Pairs object from SAM/BAM data in a single step.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sam1_file</strong> – Path to a sorted SAM/BAM file (1st mate)</p></li>
<li><p><strong>sam2_file</strong> – Path to a sorted SAM/BAM file (2nd mate)</p></li>
<li><p><strong>regions</strong> – Path to file with restriction fragments (BED, GFF)
or FASTA with genomic sequence</p></li>
<li><p><strong>restriction_enzyme</strong> – Name of restriction enzyme
(only when providing FASTA as regions)</p></li>
<li><p><strong>read_filters</strong> – List of <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to filter reads
while loading from SAM/BAM</p></li>
<li><p><strong>output_file</strong> – Path to output file</p></li>
<li><p><strong>check_sorted</strong> – Double-check that input SAM files
are sorted if True (default)</p></li>
<li><p><strong>threads</strong> – Number of threads used for finding restriction
fragments for read pairs</p></li>
<li><p><strong>batch_size</strong> – Number of read pairs sent to each restriction
fragment worker</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="fanc.pairs.generate_pairs_split">
<code class="sig-prename descclassname">fanc.pairs.</code><code class="sig-name descname">generate_pairs_split</code><span class="sig-paren">(</span><em class="sig-param">sam1_file</em>, <em class="sig-param">sam2_file</em>, <em class="sig-param">regions</em>, <em class="sig-param">restriction_enzyme=None</em>, <em class="sig-param">output_file=None</em>, <em class="sig-param">read_filters=()</em>, <em class="sig-param">check_sorted=True</em>, <em class="sig-param">threads=1</em>, <em class="sig-param">batch_size=1000000</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.pairs.generate_pairs_split" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate Pairs object from SAM/BAM files.</p>
<p>This is a convenience function that let’s you create a
Pairs object from SAM/BAM data in a single step.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sam1_file</strong> – Path to a sorted SAM/BAM file (1st mate)</p></li>
<li><p><strong>sam2_file</strong> – Path to a sorted SAM/BAM file (2nd mate)</p></li>
<li><p><strong>regions</strong> – Path to file with restriction fragments (BED, GFF)
or FASTA with genomic sequence</p></li>
<li><p><strong>restriction_enzyme</strong> – Name of restriction enzyme
(only when providing FASTA as regions)</p></li>
<li><p><strong>read_filters</strong> – List of <a class="reference internal" href="#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a> to filter reads
while loading from SAM/BAM</p></li>
<li><p><strong>output_file</strong> – Path to output file</p></li>
<li><p><strong>check_sorted</strong> – Double-check that input SAM files
are sorted if True (default)</p></li>
<li><p><strong>threads</strong> – Number of threads used for finding restriction
fragments for read pairs</p></li>
<li><p><strong>batch_size</strong> – Number of read pairs sent to each restriction
fragment worker</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="matrix.html" class="btn btn-neutral float-right" title="Matrix module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="regions.html" class="btn btn-neutral float-left" title="Regions module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Kai Kruse, Vaquerizas lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>